cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+    """Check all rate matrices are square and of appropriate dimension.
+
+    We check rates are positive in the class validator.
+    """
+    if set(qubit_dims) != set(rates):
+        raise ValueError('qubits for rates inconsistent with those through qubit_dims')
+    for q in rates:
+        if rates[q].shape != (qubit_dims[q], qubit_dims[q]):
+            raise ValueError(
+                f'Invalid shape for rate matrix: should be ({qubit_dims[q]}, {qubit_dims[q]}), '
+                f'but got {rates[q].shape}'
+            )
+
+
+@dataclass
+class ThermalNoiseModel(devices.NoiseModel):
+    """NoiseModel representing simulated thermalization of a qubit.
+
+    This model is designed for qubits which use energy levels as their states.
+    "Heating" and "cooling" here are used to refer to environmental noise which
+    transitions a qubit to higher or lower energy levels, respectively.
+    """
+
+    def __init__(
+        self,
+        qubit_dims: Dict['cirq.Qid', int],
+        gate_durations_ns: Dict[type, float],
+        heat_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        cool_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        dephase_rate_GHz: Union[float, Dict['cirq.Qid', float], None] = None,
+        require_physical_tag: bool = True,
+        skip_measurements: bool = True,
+    ):
+        """Construct a ThermalNoiseModel data object.
+
+        Required Args:
+            qubit_dims: Dimension for all qubits in the system.
+                        Currently only supports dimension=2 (qubits, not qudits)
+        Optional Args:
+            heat_rate_GHz: single number (units GHz) specifying heating rate,
+                        either per qubit, or global value for all.
+                        Given a rate gh, the Lindblad op will be sqrt(gh)*a^dag
+                        (where a is annihilation),
+                        so that the heating Lindbldian is
+                        gh(a^dag • a - 0.5{a*a^dag, •}).
+            cool_rate_GHz: single number (units GHz) specifying cooling rate,
+                        either per qubit, or global value for all.
+                        Given a rate gc, the Lindblad op will be sqrt(gc)*a
+                        so that the cooling Lindbldian is
+                        gc(a • a^dag - 0.5{n, •})
+                        This number is equivalent to 1/T1.
+            dephase_rate_GHz: single number (units GHz) specifying dephasing
+                        rate, either per qubit, or global value for all.
+                        Given a rate gd, Lindblad op will be sqrt(2*gd)*n where
+                        n = a^dag * a, so that the dephasing Lindbldian is
+                        2 * gd * (n • n - 0.5{n^2, •}).
+                        This number is equivalent to 1/Tphi.
+            require_physical_tag: whether to only apply noise to operations
+                        tagged with PHYSICAL_GATE_TAG.
```

Comment:
Docstring format seems a little off here.

Reply:
Dedented and shared lines where it was simple to do so. Trying to avoid splitting equations across lines, so I left most of this unchanged otherwise.